tests/cpp/test_frontend.cc

```diff
@@ -5,16 +5,498 @@
+  auto input_stream = rapidjson::MemoryStream(json_str.c_str(),
+                                              json_str.size());
+  std::shared_ptr<MockDelegator> delegator =
+    std::make_shared<MockDelegator>();
+  int output;
+  details::OutputHandler<int> handler {delegator, output};
+  rapidjson::Reader reader;
+
+  EXPECT_CALL(*delegator, push_delegate).Times(0);
+  EXPECT_CALL(*delegator, pop_delegate).Times(0);
+
+  rapidjson::ParseResult result = reader.Parse(input_stream, handler);
+  ASSERT_FALSE(result);
+}
+
+/******************************************************************************
+ * ArrayHandler
+ * ***************************************************************************/
+TEST(ArrayHandlerSuite, ArrayHandler) {
+  std::string json_str = "[0, 1, 2, 3]";
+  auto input_stream = rapidjson::MemoryStream(json_str.c_str(),
+                                              json_str.size());
+  std::shared_ptr<MockDelegator> delegator =
+    std::make_shared<MockDelegator>();
+
+  std::vector<int> output;
+  std::vector<int> expected{0, 1, 2, 3};
+
+  details::ArrayHandler<int> wrapped_handler {delegator, output};
+  MockArrayStarter handler {delegator, wrapped_handler};
+
+  rapidjson::Reader reader;
+  rapidjson::ParseResult result = reader.Parse(input_stream, handler);
+
+  ASSERT_TRUE(result);
+  ASSERT_EQ(output, expected);
+}
+
+/******************************************************************************
+ * TreeParamHandler
+ * ***************************************************************************/
+TEST(TreeParamHandlerSuite, TreeParamHandler) {
+  std::string json_str = "{\"num_feature\": \"1\", \"num_nodes\": \"2\","
+                         " \"size_leaf_vector\": \"3\"}";
+  auto input_stream = rapidjson::MemoryStream(json_str.c_str(),
+                                              json_str.size());
+  std::shared_ptr<MockDelegator> delegator =
+    std::make_shared<MockDelegator>();
+
+  int output;
+  int expected = 2;
+
+  details::TreeParamHandler wrapped_handler {delegator, output};
+  MockObjectStarter handler {delegator, wrapped_handler};
+
+  rapidjson::Reader reader;
+  rapidjson::ParseResult result = reader.Parse(input_stream, handler);
+
+  ASSERT_TRUE(result);
+  ASSERT_EQ(output, expected);
+}
+
+/******************************************************************************
+ * RegTreeHandler
+ * ***************************************************************************/
+TEST(RegTreeHandlerSuite, RegTreeHandler) {
+  std::string json_str = "{\"loss_changes\": []}";
+  auto input_stream = rapidjson::MemoryStream(json_str.c_str(),
+                                              json_str.size());
+  std::shared_ptr<MockDelegator> delegator =
+    std::make_shared<MockDelegator>();
+
+  Tree output;
+  details::RegTreeHandler wrapped_handler {delegator, output};
+  MockObjectStarter handler {delegator, wrapped_handler};
+
+  EXPECT_CALL(*delegator, push_delegate).Times(1);
```

Comment:
Interleaving `EXPECT_CALL()` with calls to mock functions causes an undefined behavior: https://github.com/google/googletest/issues/2828. Please move all `EXPECT_CALL` to the beginning of this unit test.

Reply:
Ooh! Good catch. I'll fix that up right away.